### DISCUSSION TITLE
v1.0.0

### DISCUSSION THREAD

**Discussion by sergiught** [auth0/terraform-provider-auth0#847] [Announcements]
#  🌍  Auth0 Terraform Provider Version 1 GA Released! ​:earth_americas:

📢  We are thrilled to announce the highly anticipated GA release of our Auth0 Terraform Provider v1! With the v1 release, managing your Auth0 resources using Terraform has never been easier. From provisioning clients, configuring connections, or setting up actions, the Auth0 Terraform Provider v1 empowers developers to streamline and automate your Auth0 infrastructure faster than ever before.

![demo-tf-generate](https://github.com/auth0/terraform-provider-auth0/assets/28300158/8bce5aaa-ad8d-4de4-b476-a9bde649fa96)

## 🔑  Key Highlights

🚀 **Auto-generating Terraform Config:** By leveraging the Auth0 CLI, you can automatically generate the necessary Terraform code to provision and manage resources in Auth0. This can save time and effort by eliminating the need to manually write and maintain Terraform configuration files.

💾  **Data Sources for Resources:** Easily fetch and reference data from existing Auth0 resources using data sources. Retrieve specific details or configuration settings of Auth0 entities and utilize them within your Terraform code, enhancing flexibility and integration.

🫂  **Resource Relationships:** Define and manage relationships between Auth0 resources effortlessly. With the ability to establish 1:1 and 1:many relationships, you can precisely model the connections and dependencies between resources, ensuring accurate provisioning and configuration.

​:closed_lock_with_key: **Zero Downtime Client Secret Rotation:** Seamlessly rotate client secrets with minimal disruption to applications and users. Includes built-in support for achieving zero downtime during client secret rotation, providing a smooth transition without impacting the availability or functionality of your applications.

🐞  **Bug Fixes and Stability Enhancements:** Auth0 Terraform Provider v1 emphasizes our dedication to delivering an enhanced, reliable tool. This version is laden with numerous improvements and comprehensive bug fixes, fortifying its stability and readiness to meet the diverse needs of the developer community.

---

Ready to get started? You can find the Auth0 Terraform Provider on the official Terraform Registry. We encourage everyone to try out the v1 and help get the word out to customers! If you are coming from a previous v0 build, be sure to check out the migration guide as this release includes breaking changes.

We also want to express our gratitude to our dedicated community and contributors for their valuable feedback and contributions that were critical to this release.

If you have any questions or need assistance, feel free to reach out here on GitHub in the Discussions thread.

Happy provisioning 🚀 

### 📚 Resources

[Quickstart](https://registry.terraform.io/providers/auth0/auth0/latest/docs/guides/quickstart) | [Documentation](https://registry.terraform.io/providers/auth0/auth0/latest/docs) | [Migration Guide](https://github.com/auth0/terraform-provider-auth0/blob/v1/MIGRATION_GUIDE.md) | [Changelog](https://github.com/auth0/terraform-provider-auth0/releases/tag/v1.0.0)

<hr /><em>This discussion was created from the release <a href='https://github.com/auth0/terraform-provider-auth0/releases/tag/v1.0.0'>v1.0.0</a>.</em>

**Comment by TakumiHaruta** [comment]
Thanks always for supporting Terraform in Auth0, and congrats on the v1 release.

I've been trying to migrate `v0.x` to `v1.0.0` in our environment according to [MIGRATION_GUIDE.md](https://github.com/auth0/terraform-provider-auth0/blob/main/MIGRATION_GUIDE.md), but I think people who use Terraform Cloud would be bothered by the change of renaming resources like `auth0_email` to `auth0_email_provider` during the upgrading because the state isn't basically contorolled via CLI in production env.

https://github.com/auth0/terraform-provider-auth0/blob/main/MIGRATION_GUIDE.md#email-provider

If you encounter an error related to renaming resources, I recommend hosting a private auth0 registry with the current `v0.x` on your Terraform Cloud before upgrading. The steps to upgrade will be like;

1. Publish private provider with the version `v0.x` on your Terraform Cloud along [Publishing Private Providers - Private Registry | Terraform | HashiCorp Developer](https://developer.hashicorp.com/terraform/cloud-docs/registry/publish-providers)
2. Switch `source` in your provider config from `auth0/auth0` to `app.terraform.io/<your-org>/auth0`
3. Apply changes to update the state file in Terraform Cloud
4. Add the `v1.x` version with a different name like `auth0new` to the provider config, and leave `app.terraform.io/<your-org>/auth0` provider as it is 
5. Add the provider meta-argument to the resources that don't have compatibilities with the `v0.x`
    - cf. [The Resource provider Meta-Argument - Configuration Language | Terraform | HashiCorp Developer](https://developer.hashicorp.com/terraform/language/meta-arguments/resource-provider)
6. Apply changes. This change will destroy `auth0_email` as well as create `auth0_email_provider`.

Looking at #731, the model itself was unchanged, so ideally it should be helpful if the old resource name `auth0_email` is still supported as `deprecated`, but here is the workaround for that

**Reply by sergiught** [reply]
Hey @TakumiHaruta 👋🏻 

Thanks for migrating to v1 and for posting the workaround for the issue you've encountered. 🙏🏻 

However you should be able to still operate the Terraform CLI even in Terraform Cloud. Have you tried following this guide: https://developer.hashicorp.com/terraform/cli/cloud ?

**Reply by TakumiHaruta** [reply]
Yeah you can, but if the workspace is integrated with the version control workflow in Terraform Cloud and the pipeline is isolated from local clients, I'd like to avoid inserting a manual step into the automated process as much as possible in production env.
